### PR TITLE
refactor: `& Impure` to `\ IO` in namespaces `Prelude`

### PR DESCRIPTION
**Pull request by ninaandrup**
Related to #4377

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze But some tests seem to fail?

**Comment by mlutze** [comment]
`println` does not have the `IO` effect, only `Impure`.

We have `IO => Impure` but we don't have `Impure => IO`.

**Comment by magnus-madsen** [comment]
> `println` does not have the `IO` effect, only `Impure`.
> 
> We have `IO => Impure` but we don't have `Impure => IO`.

To use other words, you mean that Nina should change println to have IO effect somehwere?

**Comment by mlutze** [comment]
> > `println` does not have the `IO` effect, only `Impure`.
> > We have `IO => Impure` but we don't have `Impure => IO`.
> 
> To use other words, you mean that Nina should change println to have IO effect somehwere?

Yes

**Comment by ninaandrup** [comment]
> `println` does not have the `IO` effect, only `Impure`.
> 
> We have `IO => Impure` but we don't have `Impure => IO`.

Makes sense :-)

File: main/src/library/Benchmark.flix [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace Benchmark {
     /// impure since timing is considered a side-effect in this context.
```
Comment (ninaandrup) [inline main/src/library/Benchmark.flix:20]: Should this like be `IO since timing is considered a side-effect in this context.` now ?

**Comment by ninaandrup** [comment]
Can we remember to later change this [line in `TestRedBlackTree`](https://github.com/flix/flix/blob/ee597c91a1a95dd4f0404edff7af42ae8429a921/main/test/ca/uwaterloo/flix/library/TestRedBlackTree.flix#L251) when `MutList(Static)` gets `IO` effect instead of `Impure` ? Or should/can I do something to change it in this PR?

**Comment by ninaandrup** [comment]
And should [this line not be deleted](https://github.com/flix/flix/blob/ee597c91a1a95dd4f0404edff7af42ae8429a921/main/src/library/RedBlackTree.flix#L568), since `f` has been changed to have effect `ef2` ?

File: main/src/library/Benchmark.flix (hunk 1 above)
Comment (magnus-madsen) [reply main/src/library/Benchmark.flix:20]: Yes

**Comment by magnus-madsen** [comment]
Maybe we should do prelude and some other name spaces first?

**Comment by ninaandrup** [comment]
I have run into a problem when chainging the `Prelude` namespace from `& Impure` to `\ IO`. Most of the examples have a `main` function with a `println`. But when I change `main` to
```
def main(): Unit \ IO =
    println(area(Rectangle(2, 4)))
```
in `algebraic-data-types-and-pattern-matching.flix` I get the error:
```
#
# An unexpected error has been detected by the Flix compiler:
#
#   Unable to unify: 't6994 -> Unit & Impure \ IO' and 'Unit -> Unit & Impure'.
#
# This is a bug in the Flix compiler. Please report it here:
...
```

**Comment by magnus-madsen** [comment]
Hmm. Matt? Is this because we have not yet made it default?

**Comment by magnus-madsen** [comment]
@mlutze

**Comment by mlutze** [comment]
It has to do with the `printUnlessPure` function and its assumed type in `EntryPoint`.

Since I can't really fix that without fixing the changes in this PR, I just added a commit to Nina's branch.

**Comment by magnus-madsen** [comment]
> It has to do with the `printUnlessPure` function and its assumed type in `EntryPoint`.
> 
> Since I can't really fix that without fixing the changes in this PR, I just added a commit to Nina's branch.

I am bit unsure of the steps needed to proceed.

**Comment by mlutze** [comment]
> > It has to do with the `printUnlessPure` function and its assumed type in `EntryPoint`.
> > Since I can't really fix that without fixing the changes in this PR, I just added a commit to Nina's branch.
> 
> I am bit unsure of the steps needed to proceed.

Nina should be fine to proceed as normal. I made a change to `EntryPoint.scala` which should unblock progress.

**Comment by magnus-madsen** [comment]
By changing the examples to use `\ IO` instead of `& Impure`, I presume.

**Comment by mlutze** [comment]
> By changing the examples to use `\ IO` instead of `& Impure`, I presume.

Yep

**Comment by magnus-madsen** [comment]
@mlutze Can I go ahead with this?

**Comment by mlutze** [comment]
Yes, we should be good to go. But we must be aware that once we pull the `--Xeffects` trigger, we'll break almost all client code. Community build passes because `--Xeffects` is off by default.

**Comment by magnus-madsen** [comment]
@ninaandrup You can proceed :)

**Pull request merged**